File: kmeans.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <ctype.h>
#include <float.h>

size_t rows_counter(FILE *ifp);
int vector_dim(FILE *ifp);
double distance(double *p1, double *p2, int d);

int main(int argc, char *argv[]) {
    int i,j,r,d,convergence,index,t=0;
    int k = strtol(argv[1], NULL, 10);
    int max_iter = 200;
    char *input, *output;
    double **input_vectors,**current_center,**previous_center;
    double try_min,min;

    size_t init_size;

    printf("%d", argc);
    if (argc == 5)
    {
        max_iter = strtol(argv[2], NULL, 10);
        *input = argv[3];
        *output = argv[4];
    }
    else
    {
        *input = argv[2];
        *output = argv[3];
    }

    FILE *ifp = NULL;
    FILE *ofp = NULL;


    ifp = fopen(input, "r");

    if (ifp == NULL){
        printf("An Error Has Occurred3");
        return 1;
    }

    d = vector_dim(ifp);
    init_size = rows_counter(ifp);
    if ((int)init_size < k || max_iter <= 0 || k <= 0) {
        printf("Invalid Input!");
        return 1;
    }

    /*initiate input vectors matrix*/
    input_vectors = malloc(sizeof(double*)*init_size);
    for (i = 0; i < (int)init_size; i++) {
        input_vectors[i] = malloc(sizeof(double)*d);
    }
    /*double input_vectors[init_size][d];
    initiate input vectors - insert all vectors from the input file into a matrix*/
    for (i=0; i<(int)init_size; i++)
    {
        for (j=0; j<d; j++)
        {
                fscanf(ifp, "%lf", &(input_vectors[i][j]));
                fgetc(ifp);
        }
    }
    fclose(ifp);

    /*previous centeroids initiated as the first k vectors in the input file:*/

    /*initiate matrix*/
    previous_center = malloc(sizeof(double*)*k);
    for(i = 0; i < k; i++)
    {
        previous_center[i] = malloc(sizeof(double)*d);
    }

    /*fetch first k vectors*/
    for(i = 0; i < k; i++)
    {
        for(j = 0; j < d; j++)
        {
            previous_center[i][j] = input_vectors[i][j];
        }
    }
    current_center = malloc(sizeof(double*)*k);
    for(i = 0; i < k; i++)
        current_center[i] = malloc(sizeof(double)*(d+1));

    while(t < max_iter)
    {
        /*initiate center sum matrix to sum vectors, first cell of each row is a counter*/
        double **center_sum;
        center_sum = calloc(sizeof(double*)*k, sizeof(double*));
        for(i = 0; i < k; i++)
        {
            center_sum[i] = calloc(sizeof(double)*(d+1),sizeof(double));
        }
        /* loop to find which centeroid fits which vector*/
        for(i = 0; i < (int)init_size; i++)
        {
            try_min = 0;
            min = DBL_MAX;
            index = 0;
            /*run on all centeroids for vector i*/
            for (j = 0; j < k; j++)
            {
                try_min = distance(input_vectors[i], previous_center[j], d);
                if (try_min < min)
                {
                    min = try_min;
                    index = j;
                }
            }
            /*sum the vector to the closest centeroid*/
            center_sum[index][0] += 1;
            for (r = 1; r < d + 1; r++)
            {
                center_sum[index][r] += input_vectors[i][r - 1];
            }
        }
        /*initiate current centeroids*/
        convergence = 1;
        for(i = 0; i < k; i++)
        {
            if (center_sum[i][0] == 0) {
                printf("An Error Has Occurred2");
                return 1;
            }
            for(j = 1; j < d+1; j++)
            {
                current_center[i][j-1] = (center_sum[i][j])/center_sum[i][0];
            }
            if(distance(current_center[i],previous_center[i],d) >= 0.001)
                convergence = 0;
        }

        if(convergence == 1)
            break;
        else
        {
            for(i = 0; i < k; i++)
            {
                for(j = 0; j < d; j++)
                {
                    previous_center[i][j] = current_center[i][j];
                }
            }
        }
        t ++;
    }
    ofp = fopen(output, "w");

    if (ofp == NULL){
        printf("An Error Has Occurred1");
        return 1;
    }
    
    for (i=0; i<k; i++){
        for (j=0; j<d; j++) {
            if (j == d - 1) 
            {
                fprintf(ofp, "%.4f\n",  current_center[i][j]);
            }
            else
                fprintf(ofp, "%.4f,", current_center[i][j]);
        }
    }
    fclose(ofp);
    return 0;
}



double distance(double *p1, double *p2, int d)
{   int i;
    double result = 0;
    for (i=0; i<d; i++)
    {
        result += pow((*(p1+i)-*(p2+i)), 2);
    }
    return pow(result, 0.5);
}

size_t rows_counter(FILE *ifp) {
    size_t rows = 0;
    char c;

    while (!feof(ifp)) {
        c = fgetc(ifp);
        if (c == '\n')
            rows++;
    }
    rewind(ifp);
    return rows;
}

int vector_dim(FILE *ifp){
    int d = 1;
    char c;

    do{
        c = fgetc(ifp);
        if (c == ',')
            d++;
    }

    while (c != '\n');
    rewind(ifp);
    return d;

}

File: kmeans.py
import math

def kmeans(k, input_filename, output_filename, max_iter = 200):
    if max_iter <=0 or k <=0:
        print("Invalid Input!")
        return
    input_vector_list = []

    input = open(input_filename, 'r')
    line = input.readline()
    while line != "":
        temp_vector = line.split(',')
        for index in range(len(temp_vector)):
            temp_vector[index] = float(temp_vector[index])
        input_vector_list.append(temp_vector)
        line = input.readline()
    input.close()

    if k > len(input_vector_list): # less vectors than k
        print("An Error Has Occurred")
        return
    d = len(input_vector_list[0])

    previuos_center = [input_vector_list[i] for i in range(k)]

    for iter in range(max_iter):

        current_center = [input_vector_list[i] for i in range(k)]
        cluster_sum = [[0 for i in range(d+1)] for j in range(k)]  # calculates the sum of each vector in all clusters
                                                                   # first cell is a counter
        for vector in input_vector_list: # prepping for vector cluster matching
            min =  float('inf')
            centeroid_index = 0

            for i in range(len(previuos_center)):# the loop to find which input vector goes to which cluster
                temp_distance = distance(vector, previuos_center[i])
                if temp_distance < min:
                    min = temp_distance
                    centeroid_index = i

            cluster_sum[centeroid_index][0] += 1 # the cluster which a vector was matched with gets +1 counter
            for i in range(d):
                cluster_sum[centeroid_index][i+1]+= vector[i] # adds that vector's value to the centeroid

        convergence = [0 for i in range(k)] # array for delta mew's convergence
        for i in range(len(cluster_sum)): # update the centeroids
            cluster_size = cluster_sum[i][0]
            if(cluster_size == 0):
                print("An Error Has Occurred")
                return
            current_center[i] = [cluster_sum[i][j]/cluster_size for j in range(1,d+1)]
            if distance(current_center[i],previuos_center[i]) < 0.001:
                convergence[i] = 1

        if sum(convergence) == k:
            break
        else:
            previuos_center = current_center

    output = open(output_filename, 'w')
    for i in range(len(current_center)):
        current_center[i] = [str("%.4f" % round(current_center[i][j], 4)) for j in range(d)]
        line = ','.join(current_center[i])
        output.write(line + '\n')
    output.close()

def distance(x , y):
    result = 0
    for i in range(len(x)):
        result += (x[i]-y[i])**2
    return (result**0.5)